zenoh-flow/src/io/input.rs

```diff
@@ -0,0 +1,190 @@
+//
+// Copyright (c) 2022 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use crate::prelude::{ErrorKind, Message, PortId, ZFData};
+use crate::types::{Data, DataMessage, LinkMessage};
+use crate::{zferror, Result};
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use uhlc::Timestamp;
+
+/// The [`Inputs`](`Inputs`) structure contains all the receiving channels we created for a
+/// [`Sink`](`Sink`) or an [`Operator`](`Operator`).
+///
+/// To access these underlying channels, two methods are available:
+/// - `take`: this will return an `Input<T>` where `T` implements [`ZFData`](`ZFData`),
+/// - `take_raw`: this will return an [`InputRaw`](`InputRaw`) — a type agnostic receiver.
+///
+/// Choosing between `take` and `take_raw` is a trade-off between convenience and performance: an
+/// `Input<T>` conveniently receives instances of `T` and is thus less performant as Zenoh-Flow has
+/// to manipulate the data; an `InputRaw` is more performant but all the manipulation must be
+/// performed in the code of the Node. An `InputRaw` can be leveraged, for instance, for
+/// load-balancing or rate-limiting.
+pub struct Inputs {
+    pub(crate) hmap: HashMap<PortId, Vec<flume::Receiver<LinkMessage>>>,
+}
+
+// Dereferencing on the internal [`HashMap`](`Hashmap`) allows users to call all the methods
+// implemented on it: `keys()` for one.
+impl Deref for Inputs {
+    type Target = HashMap<PortId, Vec<flume::Receiver<LinkMessage>>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.hmap
+    }
+}
+
+impl Inputs {
+    pub(crate) fn new() -> Self {
+        Self {
+            hmap: HashMap::default(),
+        }
+    }
+
+    /// Insert the `flume::Receiver` in the [`Inputs`](`Inputs`), creating the entry if needed in
+    /// the internal `HashMap`.
+    pub(crate) fn insert(&mut self, port_id: PortId, rx: flume::Receiver<LinkMessage>) {
+        self.hmap
+            .entry(port_id)
+            .or_insert_with(Vec::default)
+            .push(rx)
+    }
+
+    /// Returns the typed [`Input<T>`](`Input`) associated to the provided `port_id`, if one is
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Performance
+    ///
+    /// With a typed [`Input<T>`](`Input`), Zenoh-Flow will perform operations on the underlying
```

Comment:
We could add that deserialization is done only if the inner data is serialized (i.e., coming from the network or the upstream node sent it as a byte array.

Reply:
The text in the line after isn't explicit enough?

> if the data are received serialized then Zenoh-Flow will deserialize them, if they are received "typed" then Zenoh-Flow will check that the type matches what is expected.

Reply:
I missed it :)